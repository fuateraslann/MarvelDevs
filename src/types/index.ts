export type TImageExtensions = 'jpg' | 'png' | 'svg'
export type TCharacter = {
  id: number
  name: string
  description: string
  thumbnail: {
    path: string
    extension: TImageExtensions
  }
}
export type TComics = {
  id: number
  title: string
  thumbnail: {
    path: string
    extension: TImageExtensions
  }
}
