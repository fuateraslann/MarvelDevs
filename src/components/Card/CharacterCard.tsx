import { Button } from '@mui/material'
import DetailsModal from 'components/DetailsModal'
import { useState } from 'react'
import { TCharacter } from 'types'
import Card from '.'
const CharacterCard = ({ character }: { character: TCharacter }) => {
  const [openModal, setOpenModal] = useState(false)
  const handleOpenModal = () => {
    setOpenModal(true)
  }

  const handleCloseModal = () => {
    setOpenModal(false)
  }

  return (
    <Card
      name={character.name}
      imgSrc={character.thumbnail.path + '.' + character.thumbnail.extension}
      extras={
        <>
          <Button
            sx={{
              maxHeight: 30,
              border: '1px solid',
              borderRadius: 1,
            }}
            onClick={handleOpenModal}
          >
            Details
          </Button>
          <DetailsModal
            detailsItem={character}
            openModal={openModal}
            onCloseModal={handleCloseModal}
          />
        </>
      }
    />
  )
}

export default CharacterCard
