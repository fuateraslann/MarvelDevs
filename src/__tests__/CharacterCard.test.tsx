/* eslint-disable react/display-name */
import React from 'react'
import { render, fireEvent } from '@testing-library/react'
import '@testing-library/jest-dom/extend-expect'
import CharacterCard from 'components/Card/CharacterCard'
import { TImageExtensions } from 'types'

jest.mock(
  'components/DetailsModal',
  () =>
    ({ openModal }: { openModal: boolean }) =>
      <div>{openModal ? <div>DetailsMock</div> : null}</div>
)
jest.mock(
  'components/Card',
  () =>
    ({
      name,
      imgSrc,
      extras,
    }: {
      name: string
      imgSrc: string
      extras: React.ReactNode
    }) =>
      (
        <div>
          <div>{name}</div>
          <div>{imgSrc}</div>
          <div>{extras}</div>
        </div>
      )
)

const mockCharacter = {
  id: 1,
  description: 'description of character',
  name: 'Test Character',
  thumbnail: {
    path: 'test/path',
    extension: 'jpg' as TImageExtensions,
  },
}
describe('CharacterCard', () => {
  test('renders CharacterCard correctly', () => {
    const { asFragment, getByText } = render(
      <CharacterCard character={mockCharacter} />
    )
    expect(getByText('Test Character')).toBeInTheDocument()
    expect(getByText('Details')).toBeInTheDocument()
    expect(asFragment()).toMatchSnapshot() // Snapshot test
  })

  test('opens DetailsModal when "Details" button is clicked', () => {
    const { getByText, asFragment } = render(
      <CharacterCard character={mockCharacter} />
    )
    fireEvent.click(getByText('Details'))
    expect(getByText('DetailsMock')).toBeInTheDocument()
    expect(asFragment()).toMatchSnapshot() // Snapshot test
  })
})
