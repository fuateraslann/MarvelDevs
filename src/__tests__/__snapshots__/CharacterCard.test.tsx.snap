// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CharacterCard opens DetailsModal when "Details" button is clicked 1`] = `
<DocumentFragment>
  <div>
    <div>
      Test Character
    </div>
    <div>
      test/path.jpg
    </div>
    <div>
      <button
        class="MuiButtonBase-root MuiButton-root MuiButton-text MuiButton-textPrimary MuiButton-sizeMedium MuiButton-textSizeMedium MuiButton-root MuiButton-text MuiButton-textPrimary MuiButton-sizeMedium MuiButton-textSizeMedium css-1c8yy18-MuiButtonBase-root-MuiButton-root"
        tabindex="0"
        type="button"
      >
        Details
        <span
          class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
        />
      </button>
      <div>
        <div>
          DetailsMock
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;

exports[`CharacterCard renders CharacterCard correctly 1`] = `
<DocumentFragment>
  <div>
    <div>
      Test Character
    </div>
    <div>
      test/path.jpg
    </div>
    <div>
      <button
        class="MuiButtonBase-root MuiButton-root MuiButton-text MuiButton-textPrimary MuiButton-sizeMedium MuiButton-textSizeMedium MuiButton-root MuiButton-text MuiButton-textPrimary MuiButton-sizeMedium MuiButton-textSizeMedium css-1c8yy18-MuiButtonBase-root-MuiButton-root"
        tabindex="0"
        type="button"
      >
        Details
        <span
          class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
        />
      </button>
      <div />
    </div>
  </div>
</DocumentFragment>
`;
